bot/cogs/extensions.py

```diff
@@ -0,0 +1,216 @@
+    UNLOAD = functools.partial(Bot.unload_extension)
+    RELOAD = functools.partial(Bot.reload_extension)
+
+
+class Extension(commands.Converter):
+    """
+    Fully qualify the name of an extension and ensure it exists.
+
+    The * and ** values bypass this when used with the reload command.
+    """
+
+    async def convert(self, ctx: Context, argument: str) -> str:
+        """Fully qualify the name of an extension and ensure it exists."""
+        # Special values to reload all extensions
+        if ctx.command.name == "reload" and (argument == "*" or argument == "**"):
+            return argument
+
+        argument = argument.lower()
+
+        if "." not in argument:
+            argument = f"bot.cogs.{argument}"
+
+        if argument in EXTENSIONS:
+            return argument
+        else:
+            raise commands.BadArgument(f":x: Could not find the extension `{argument}`.")
+
+
+class Extensions(commands.Cog):
+    """Extension management commands."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @group(name="extensions", aliases=("ext", "exts", "c", "cogs"), invoke_without_command=True)
+    async def extensions_group(self, ctx: Context) -> None:
+        """Load, unload, reload, and list loaded extensions."""
+        await ctx.invoke(self.bot.get_command("help"), "extensions")
+
+    @extensions_group.command(name="load", aliases=("l",))
+    async def load_command(self, ctx: Context, extension: Extension) -> None:
+        """Load an extension given its fully qualified or unqualified name."""
+        msg, _ = self.manage(extension, Action.LOAD)
+        await ctx.send(msg)
+
+    @extensions_group.command(name="unload", aliases=("ul",))
+    async def unload_command(self, ctx: Context, extension: Extension) -> None:
+        """Unload a currently loaded extension given its fully qualified or unqualified name."""
+        if extension in UNLOAD_BLACKLIST:
+            msg = f":x: The extension `{extension}` may not be unloaded."
+        else:
+            msg, _ = self.manage(extension, Action.UNLOAD)
+
+        await ctx.send(msg)
+
+    @extensions_group.command(name="reload", aliases=("r",))
+    async def reload_command(self, ctx: Context, *extensions: Extension) -> None:
+        """
+        Reload extensions given their fully qualified or unqualified names.
+
+        If an extension fails to be reloaded, it will be rolled-back to the prior working state.
+
+        If `*` is given as the name, all currently loaded extensions will be reloaded.
+        If `**` is given as the name, all extensions, including unloaded ones, will be reloaded.
+        """
+        if not extensions:
+            await ctx.invoke(self.bot.get_command("help"), "extensions reload")
+            return
+
+        if len(extensions) > 1:
+            msg = await self.batch_reload(*extensions)
+        else:
+            msg, _ = self.manage(extensions[0], Action.RELOAD)
```

Comment:
The docstring leads me to believe that providing a `*` or `**` as the single argument to this subcommand would respectively reload all the loaded cogs or all of the cogs regardless of their status. However, this `if-else` block will call the `self.manage` method directly when a single argument is provided. Since `self.manage` does not have any logic to handle asterisk characters, it will fail with:

```py
Oct 07 09:35:01 Bot: |               bot.cogs.extensions |    ERROR | Extension '*' failed to load.
Traceback (most recent call last):
  File "/home/sebastiaan/pydis/repositories/djangobot/bot/cogs/extensions.py", line 180, in manage
    action.value(self.bot, ext)
  File "/home/sebastiaan/.local/share/virtualenvs/djangobot-Py5o4YeS/lib/python3.7/site-packages/discord/ext/commands/bot.py", line 687, in reload_extension
    raise errors.ExtensionNotLoaded(name)
discord.ext.commands.errors.ExtensionNotLoaded: Extension '*' has not been loaded.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sebastiaan/.local/share/virtualenvs/djangobot-Py5o4YeS/lib/python3.7/site-packages/discord/ext/commands/bot.py", line 621, in load_extension
    lib = importlib.import_module(name)
  File "/home/sebastiaan/.local/share/virtualenvs/djangobot-Py5o4YeS/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named '*'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sebastiaan/pydis/repositories/djangobot/bot/cogs/extensions.py", line 180, in manage
    action.value(self.bot, ext)
  File "/home/sebastiaan/.local/share/virtualenvs/djangobot-Py5o4YeS/lib/python3.7/site-packages/discord/ext/commands/bot.py", line 623, in load_extension
    raise errors.ExtensionNotFound(name, e) from e
discord.ext.commands.errors.ExtensionNotFound: Extension '*' could not be loaded.
```

That means that, effectively, we can only use `*`/`**` in combination with at least one other argument. I don't think that's very intuitive. I think it's better for asterisk arguments to work on their own as well.

Reply:
It was my intention for it to work on its own. I broke it when I was attempting to simplify the logic here. I think the fix is

```py
if len(extensions) > 1 or "*" in extensions:
```

on line 99. I'll test this tomorrow.